商品出品機能の実装

jaxzzxaj opened the pull request:
WHAT
商品出品機能の実装

WHY
userが新規登録やログイン後に商品を出品ができるようにするため

jaxzzxaj committed:
出品の実装１

・routesでcreateの追加
・controllerの修正とcreate設定
・viewのhogeをvalidationに合わせる
・viewの画像添付のため、gemの導入
・active storage のDBを作成した後、imageをproductとつなげる

jaxzzxaj committed:
activehashでプルダウンの実装等

jaxzzxaj committed:
価格を固定

jaxzzxaj committed:
validatesの修正

一個のvalidatesに複数のカラムを記述しても良い！

jaxzzxaj committed:
データを保存するための修正

genreID必要なし
validationで数字のみで良い（半角英字はなしと書かなくともPCは通じるから）

jaxzzxaj commented:
ログインしてから出品画面へ移動する動画

https://gyazo.com/b71b5c76441b30480e9d2b7b0bd03fb0

jaxzzxaj commented:
金額を入力したあと、販売手数料と販売利益を表示する動画

https://gyazo.com/72796fef9ada72de86225d2b63b83e52

jaxzzxaj commented:
正しく出品内容を記入すると商品が出品できる動画
（必要な情報を適切に入力すると、商品情報がデータベースに保存されること）

https://gyazo.com/f7224a63e8ed4c05add2e9cd710284ee

jaxzzxaj commented:
入力間違えがある場合、エラーメッセージがでるようになる

https://gyazo.com/e30e0910f07d920089f6278dbe9077d4

jaxzzxaj committed:
未ログイン者の強制遷移

ログインページへ遷移するの修正

jaxzzxaj commented:
未ログイン者が出品ボタンをを押してしまったときログインページへ遷移させる動画

https://gyazo.com/1da7b854aad77487ded686fa44783b3b

jaxzzxaj committed:
テストコード

jaxzzxaj commented:
テストコードの画像

https://gyazo.com/31415889d0183eaa807042f23a4b4f40

shodai-ito commented on app/controllers/products_controller.rb:
```diff
@@ -1,25 +1,43 @@
-  # def index
-  #  @products = Product.includes(:user)
-  # end
+  before_action :move_to_index, except: :index
```
`before_action :move_to_index, except: :index`
authenticate_user!の記述があるので、こちらの記述は不要です！

shodai-ito commented on app/controllers/products_controller.rb:
```diff
@@ -1,25 +1,43 @@
-  # end
+  before_action :move_to_index, except: :index
+  #下記の記述で@productsが新規投稿順に並ぶようにしている。
+  def index
+    @product = Product.includes(:user)
+    @products = Product.order("ceated_at DESK")
+  end
```
こちら次回の実装内容になりますので、いったんコメントアウトをお願いします！

shodai-ito commented on app/controllers/products_controller.rb:
```diff
@@ -1,25 +1,43 @@
+    params.require(:product).permit(:name,:category_id,:product_condition_id,:shipping_charges_id,:estimated_shipping_date_id,:prefecture_id,:price,:text,:image).merge(user_id: current_user.id)
+  end
+
+  def move_to_index
+    unless user_signed_in?
+      redirect_to action: :index
+    end
   end
```
こちらのメソッドは不要です！

shodai-ito commented on app/models/product.rb:
```diff
@@ -1,14 +1,20 @@
+  end
+
+  with_options numericality:{ only_integer: true, greater_than: 300, less_than: 9999999} do
+    validates :price,format: {with: /\A[a-zA-Z0-9]+\z/,message: "is invalid.Input half-width alphanumeric."}
```
`/\A[a-zA-Z0-9]+\z/`
こちらの正規表現ですとローマ字も許容してしまいます。
修正しましょう！

shodai-ito commented on app/views/products/new.html.erb:
```diff
@@ -0,0 +1,161 @@
+    <%# メモ：金額設定するからlocaltruedoにならない気がする %>
+    <%= form_with model: @product, url:products_path,local: true do |f| %>
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: f.object %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
実装が完了したコメントアウトは削除しておきましょう！

shodai-ito commented on spec/factories/product.rb:
```diff
@@ -0,0 +1,22 @@
+FactoryBot.define do
+  factory :product do
+
+    shipping_charges_id        {'1'}
+    product_condition_id       {'1'}
+    #activestorageをつかってるから記述方法を訂正。→oooo.jpは不可。
+    # image                      {"moon.jp"}
+    prefecture_id              {'1'}
+    estimated_shipping_date_id {'2'}
+    category_id                {'4'}
+    name                       {"月"}
+    text                       {"月の絵"}
+    price                      {'500'}
```
integer型のダミーデータは''を外しましょう！

shodai-ito commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,131 @@
+        expect(@product).to be_valid
+      end
+
+      it "価格は300以上9999999以下であり、半角数字のみで入力する" do
+        @product.price           = '599'
+      end
```
こちらのテストコードは現状機能していないため削除してしまいましょう！

jaxzzxaj replied:
小池一輝メンターさんに聞いてみたところ

基本的に@product.valid?がないとテストしたことにならない。自分が書いたコードは@product.priceに599を代入しただけな状態なのでエラーが出るわけがない、なのでターミナルで緑になる。でも、300以上9999999以下であることは立証されているからooooo.valid?は書かなくとも良い、このコード自体いらないですよと教えてくれたのですが、、

modelのvalidationは基本的にすべてテストしなければならないと認識しているのですが、この300以上9999999以下の設定はテストをしなくともいいのでしょうか？

yutomentors replied:
ご質問ありがとうございます！

記述いただいている正常系のテストコードですが、**本コード自体なくとも今回の実装としては問題ございません。**
`300以上9999999以下`のテストは異常系のテストの際に記述いただければと思いす。
既に`299以下だと不可``999999999以上だと不可` についてテストを実行いただいていますのでこちらについては問題ございません。

shodai-ito commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,131 @@
+        expect(@product.errors.full_messages).to include("Category can't be blank")
+      end
+
+      it "category_idが0だと不可" do
+        @product.category_id = '0'
+        @product.valid?
+        expect(@product.errors.full_messages).to include("Category must be other than 0")
+      end
```
integer型の値を代入したいので''は外しましょう！
他のactive_hashの項目についても同様です！

shodai-ito commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,131 @@
+          expect(@product.errors.full_messages).to include("Price can't be blank")
+        end
+
+        it "priceが299以下だと不可" do
+          @product.price = '100'
+          @product.valid?
+          expect(@product.errors.full_messages).to include("Price must be greater than 300")
+        end
+
+        it "priceが10000000以上だと不可" do
+          @product.price = '10000000'
+          @product.valid?
+          expect(@product.errors.full_messages).to include("Price must be less than 9999999")
+        end
```
integer型のデータを代入する際は''を外しましょう！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

その後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

jaxzzxaj committed:
未ログイン者をログインページへ遷移

authenticateがあるのでmove_to_indexは必要ない

jaxzzxaj committed:
次の実装のものなのでコメントアウトをする

jaxzzxaj committed:
privateのmove_to_indexも不要

jaxzzxaj committed:
priceの正規表現飲み直し

ローマ字も→数字のみ

jaxzzxaj committed:
コメントアウトを削除

実装が終わったため確認のためのエラーメッセージはもう必要ない

jaxzzxaj committed:
文字列と数字の扱い方の違い

数字はダブルクォーテーションはひつようない。

jaxzzxaj committed:
数字ではなく文字列なため再修正

yutomentors commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,131 @@
+        expect(@product).to be_valid
+      end
+
+      it "価格は300以上9999999以下であり、半角数字のみで入力する" do
```
```ruby
it "価格は300以上9999999以下であり、半角数字のみで入力する" do
@product.price = 599
end
```

今回正常系においてこちらの項目はなくて大丈夫でございます！
※上の正常系の記述にて全ての項目が入力されているときにテストが正しく通過しているためです。
金額だけ重複して正常系テストを行う必要性はないため今回は不要です！

また、異常系のテストコードの箇所にてきちんと`300以下だと不可``999999999より上だと不可`の検証ができているため問題ございません！

yutomentors reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

jaxzzxaj committed:
テストコードの修正

異常系のテストコードにもあるから不必要

kuu0123 commented on app/controllers/products_controller.rb:
```diff
@@ -1,25 +1,38 @@
-  #     redirect_to action: :index
-  #   end
-  # end
+  def user_params
```
```ruby
def user_params
```
メソッド名は、できる限りメソッド内での処理がわかる命名が好ましいですね。
今回であれば、出品した商品の情報を受け取るストロングパラメータになるため、
`user`ではなく、`product`のparamsとする方がより好ましい命名になりそうです。

kuu0123 commented on app/models/product.rb:
```diff
@@ -1,14 +1,20 @@
-  #belongs_to :user
+  belongs_to :user
+  has_one_attached :image
+  extend ActiveHash::Associations::ActiveRecordExtensions
```
```ruby
extend ActiveHash::Associations::ActiveRecordExtensions
```
上記の記述はありますが、商品のモデルと各active hashのモデルのアソシエーションは記述されていないようです。
カリキュラムを参考にして、追加しましょう。

kuu0123 commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,126 @@
+          expect(@product.errors.full_messages).to include("Price must be less than 9999999")
+        end
+
+        it "priceが全角数字だと不可" do
+          @product.price = '１０００'
+          @product.valid?
+          expect(@product.errors.full_messages).to include("Price is not a number")
+        end
+
+        it "priceが英字だと不可" do
+          @product.price = "abcdef"
+          @product.valid?
+          expect(@product.errors.full_messages).to include("Price is not a number")
+        end
```
「priceが半角数字でなければ保存できないこと」を確かめるため、以下3点は分けて考える必要があります。
不足しているテストを追加で実装しましょう！
```
・全角文字では登録できないこと
・半角英数混合では登録できないこと  ## 追加しましょう！
・半角英語だけでは登録できないこと
```

kuu0123 reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加で3点コメントいたしましたので、ご確認お願いします！

jaxzzxaj committed:
paramsの修正

わかりやすくproductに変更

jaxzzxaj committed:
end忘れ

修正したときにendも一緒に修正してしまった。

jaxzzxaj commented:
テストコードに”英数混合だと不可を追加”verのgyazo

https://gyazo.com/6a57c95795e0b7c434146678978c85a0

jaxzzxaj committed:
productとactivehashを紐付ける

kazutaka-div commented on db/migrate/20210402081556_create_products.rb:
```diff
@@ -9,7 +9,6 @@ def change
       t.string       :prefecture_id,              null: false
       t.integer      :price,                      null: false
       t.text         :text,                       null: false
-      #t.references   :user,                       foreign_key:true
```
```ruby
 #t.references   :user,                       foreign_key:true
```
上記、削除されていますが、商品のテーブルには、「どのユーザーが出品したか」のidを保存する必要がございます。
そのため、上記の記述はテーブルに戻すよう修正をお願いいたします。

kazutaka-div reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認の上、
再度、レビュー依頼フォームをご提出ください。

jaxzzxaj committed:
user_idを一つのファイルに収める

・add_referenceのファイルを消してテーブルの作成順番を変える。

Katsuno0814 commented on app/models/product.rb:
```diff
@@ -1,14 +1,21 @@
+    validates :category_id, :product_condition_id, :shipping_charges_id, :estimated_shipping_date_id, :prefecture_id
+  end
+
+  with_options numericality:{ only_integer: true, greater_than: 300, less_than: 9999999} do
```
>with_options numericality:{ only_integer: true, greater_than: 300, less_than: 9999999} do

こちらgreater_than: 299, less_than:10000000と修正いたしましょう！
現状ですと300、9999999円で出品できなくなります！

jaxzzxaj replied:
未だに不等号？が苦手でした。
以上(greater_than)は＜。以下（lesser）は＞。

Katsuno0814 reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いします！

修正頂いた後、コミットプッシュを確認いただき、再度レビュー依頼のご提出をお願い致します！

jaxzzxaj committed:
出品の値段修正

300→299
9999999→10000000

Hide-Morishita commented on app/models/product.rb:
```diff
@@ -1,14 +1,21 @@
+  belongs_to :user
+  has_one_attached :image
+  extend ActiveHash::Associations::ActiveRecordExtensions
+  belongs_to :category, :product_condition, :shipping_charges, :estimated_shipping_date, :prefecture
```
```ruby
belongs_to :category, :product_condition, :shipping_charges, :estimated_shipping_date, :prefecture
```
アソシーションの記述は、まとめて記述すると
正常に紐付けができない可能性がございますので、
下記のように、紐付けたいモデルごとに分けて記述しましょう。
```ruby
belongs_to :category
belongs_to :product_condition
belongs_to :shipping_charges
belongs_to :estimated_shipping_date
belongs_to :prefecture
```

jaxzzxaj replied:
rubyのルールで何度も同じものを書くな！というのがあったと思ってアソシエーションの短縮化？を探してみたのですが見つからず諦めました。

Hide-Morishita reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加で**1点**コメントいたしましたので、ご確認お願いいたします。
コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りいただくようお願いいたします。

jaxzzxaj committed:
belongs_to の修正

yutamasumoto commented on spec/models/product_spec.rb:
```diff
@@ -0,0 +1,133 @@
+          @product.price = "abcdef"
+          @product.valid?
+          expect(@product.errors.full_messages).to include("Price is not a number")
+        end
```
現状ですとuserの有無に対してのテストが実装されておりませんので下記内容を追記しましょう。
```
     It userが存在しないと登録できない do
```

yutamasumoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

jaxzzxaj committed:
テストコードの修正

テストするときの書き方２種類

・factory(:user)
をcreateしたら、user_idが生成される

・factory(:user)をbuiltしたら、userが生成される

・上記を踏まえて、factory(:user）と(:product)をassociationでつないだら、カラムの指定に注意する！

今回は後者なので、@product.user = nilで大丈夫！

chobi555 commented:
レビュー依頼ありがとうございます！
LGTMとさせていただきます、実装お疲れさまでした！

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き頑張っていきましょう！
